src/lib.rs

```diff
@@ -326,19 +352,34 @@ pub trait BatchDomain: Domain {
 }
 
 /// Trait that provides access to the components of an encrypted transaction output.
-///
-/// Implementations of this trait are required to define the length of their ciphertext
-/// field. In order to use the trial decryption APIs in this crate, the length must be
-/// either [`ENC_CIPHERTEXT_SIZE`] or [`COMPACT_NOTE_SIZE`].
-pub trait ShieldedOutput<D: Domain, const CIPHERTEXT_SIZE: usize> {
+pub trait ShieldedOutput<D: Domain> {
     /// Exposes the `ephemeral_key` field of the output.
     fn ephemeral_key(&self) -> EphemeralKeyBytes;
 
     /// Exposes the `cmu_bytes` or `cmx_bytes` field of the output.
     fn cmstar_bytes(&self) -> D::ExtractedCommitmentBytes;
 
-    /// Exposes the note ciphertext of the output.
-    fn enc_ciphertext(&self) -> &[u8; CIPHERTEXT_SIZE];
+    /// Exposes the note ciphertext of the output. Returns `None` if the output is compact.
+    fn enc_ciphertext(&self) -> Option<&D::NoteCiphertextBytes>;
+
+    // FIXME: Should we return `Option<D::CompactNoteCiphertextBytes>` or
+    // `&D::CompactNoteCiphertextBytes` instead? (complexity)?
+    /// Exposes the compact note ciphertext of the output.
+    fn enc_ciphertext_compact(&self) -> D::CompactNoteCiphertextBytes;
+
+    //// Splits the AEAD tag from the ciphertext.
+    fn split_ciphertext_at_tag(&self) -> Option<(D::NotePlaintextBytes, [u8; AEAD_TAG_SIZE])> {
+        let enc_ciphertext_bytes = self.enc_ciphertext()?.as_ref();
+
+        let (plaintext, tail) = enc_ciphertext_bytes
+            .len()
+            .checked_sub(AEAD_TAG_SIZE)
+            .map(|tag_loc| enc_ciphertext_bytes.split_at(tag_loc))?;
+
+        let tag: [u8; AEAD_TAG_SIZE] = tail.try_into().expect("the length of the tag is correct");
+
+        D::parse_note_plaintext_bytes(plaintext).map(|plaintext| (plaintext, tag))
```

Comment:
Here also I think a panic would be correct. Inside `D::parse_note_plaintext_bytes` we don't know if `plaintext` is the correct length, so it needs to be fallible. Here however, we know we obtained `plaintext` from a `D::NoteCiphertextBytes`, and it would therefore be a programming error if that type was inconsistent with `D::NotePlaintextBytes`. The only case fallibility would matter here is if `D::parse_note_plaintext_bytes` did some type-checking, but AFAIK it is only checking the length, and the actual parsing into protocol-specific types happens in a separate method.

After that change, `None` would now just mean "this type has no full note ciphertext".
```suggestion
        Some((
            D::parse_note_plaintext_bytes(plaintext)
                .expect("D::NoteCiphertextBytes and D::NotePlaintextBytes should be consistent"),
            tag,
        ))
```